### PR TITLE
LOOP-1201 Export of Critical Event Logs

### PR DESCRIPTION
**Pull request by darinkrauss**
- https://tidepool.atlassian.net/browse/LOOP-1201
- Replace StoredSettings.InsulinModel with StoredInsulinModel
- Implement Codable StoredDosingDecision.NotificationSettings
- Add ZipArchive
- Add CriticalEventLogExportManager
- Add CriticalEventLogExportView and CriticalEventLogExportViewModel
- Add Checkmark image for successful critical event log export
- Add minizip dependency

Note: This will require @ps2 and one other reviewer. Also, I looked into breaking this into smaller PRs, but could not come up with a way to do so that made any sense. (I could have broken out the `Codable` implementations, but that doesn't make this PR any easier, just a bit smaller, and is not worth the work.)

Note: All of the store implementations of `CriticalEventLog` are very similar (with only a few differences unique to each store). Once you've understood one, you've understood 95% of the rest).

### CODE REVIEW

**Comment by darinkrauss** [comment]
Related LoopWorkspace PR for demonstration of build: https://github.com/tidepool-org/LoopWorkspace/pull/250

File: Cartfile [hunk 1]
```diff
@@ -6,3 +6,4 @@ github "LoopKit/G4ShareSpy" "merge-from-tidepool"
 github "ps2/rileylink_ios" "merge-from-tidepool"
 github "LoopKit/Amplitude-iOS" "decreepify"
 github "LoopKit/MKRingProgressView" "appex-safe"
+github "dexman/Minizip" "master"
```
Comment (rickpasetto) [inline Cartfile:9]: Hm.  This dependency has not been touched since 2018.  Are you sure it's legit?

File: Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift [hunk 2]
```diff
@@ -124,6 +124,18 @@ fileprivate extension StoredDosingDecision {
                                                   pumpBatteryChargeRemaining: 3.5,
                                                   basalDeliveryState: .initiatingTempBasal,
                                                   bolusState: .none)
+        let notificationSettings = StoredDosingDecision.NotificationSettings(authorizationStatus: .authorized,
```
Comment (rickpasetto) [inline Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift]: I'd love to know what this change is about...care to help me understand?

File: Loop/Extensions/OutputStream.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//  Loop
+//
+//  Created by Darin Krauss on 8/28/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+extension OutputStream {
+    func write(_ string: String) throws {
```
Comment (rickpasetto) [inline Loop/Extensions/OutputStream.swift:12]: [nit] unit tests for these?

File: Loop/Managers/Alerts/AlertStore.swift [hunk 4]
```diff
@@ -349,6 +349,88 @@ extension Result where Success == Void {
...
+
+        if let closeError = encoder.close(), error == nil {
+            error = closeError
+        }
+
+        return error
+    }
+
+    private func exportDatePredicate(startDate: Date, endDate: Date? = nil) -> NSPredicate {
+        var issuedDatePredicate = NSPredicate(format: "issuedDate >= %@", startDate as NSDate)
```
Comment (rickpasetto) [inline Loop/Managers/Alerts/AlertStore.swift:422]: What is the requirement with respect to how `startDate` applies to Alerts?
Seems odd to me to export checking all three of these dates.  What is the reasoning behind this?
I would have expected only checking `issuedDate` (which will always be the earliest of the dates)

Also...this doesn't take into account the delay in the Alert.  So...an alert that was _issued_ but not yet _shown_ will appear in the logs.

Also...are you aware that alert entries that are retracted before they are shown are _deleted_ from the database?  

I think I would like to pow-wow with you about what the expectations are around `startDate`

File: Loop/Managers/DeviceDataManager.swift [hunk 5]
```diff
@@ -34,8 +35,8 @@ final class DeviceDataManager {
     /// The last time a BLE heartbeat was received and acted upon.
     private var lastBLEDrivenUpdate = Date.distantPast
 
-    private var deviceLog: PersistentDeviceLog
-    
+    var deviceLog: PersistentDeviceLog
```
Comment (rickpasetto) [inline Loop/Managers/DeviceDataManager.swift]: Hm.  I don't see where you need to change the protection level of this.  Can this be reverted?

File: Loop/Managers/DeviceDataManager.swift [hunk 6]
```diff
@@ -992,6 +1000,77 @@ extension DeviceDataManager {
...
+        } catch let error {
+            #if IOS_SIMULATOR
+            log.debug("Failed to schedule critical event log export background task due to running on simulator")
+            #else
+            log.error("Failed to schedule critical event log export background task: %{public}@", String(describing: error))
+            #endif
+        }
+    }
+
+    public func removeExportsDirectory() -> Error? {
```
Comment (rickpasetto) [inline Loop/Managers/DeviceDataManager.swift:1057]: [nit] consider just having this function throw and let users catch the Error (if they really just want to ignore the error, they can always just call it with `try?`).

File: Loop/Managers/ZipArchive.swift [hunk 7]
```diff
@@ -0,0 +1,274 @@
+//
+//  ZipArchive.swift
```
Comment (rickpasetto) [inline Loop/Managers/ZipArchive.swift]: [nit] Should this be in `Common` (maybe in a `Utilities` subgroup?) instead of `Managers`?

File: Loop/Managers/ZipArchive.swift [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                return false
+            }
+            status = to
+            return true
+        }
+
+        @discardableResult
+        private func checkZipError(_ err: Int32) -> Error? {
+            if err != ZIP_OK {
+                setError(ZipArchiveError.internalFailure(err))
```
Comment (rickpasetto) [inline Loop/Managers/ZipArchive.swift]: should this `return setError(...)`?

File: Loop/Managers/ZipArchive.swift [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+        // MARK: - Internal
+
+        fileprivate func unlockedClose() {
+            transitionStatus(from: .open, to: .closed)
+            checkZipError(zipCloseFileInZip(archive.file))
+            archive.stream = nil
+        }
+
+        @discardableResult
```
Comment (rickpasetto) [inline Loop/Managers/ZipArchive.swift]: [nit] I have to confess I'm having a hard time following the error handling here.  Perhaps consider `throw`ing errors instead?  It would avoid having to return and propagate errors, avoids the `@discardableResult`, and should make the calling code a bit more readable.  Not blocking.

File: Loop/Managers/CriticalEventLogExportManager.swift [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+    /// Has the export been cancelled?
+    var isCancelled: Bool { get }
+
+    /// Cancel the export progress.
+    func cancel()
+
+    /// Export
+    /// - Parameter now: - The current time.
+    /// - Returns: Any error that occurred during export, including cancelled.
+    func export(now: Date) -> Error?
```
Comment (rickpasetto) [inline Loop/Managers/CriticalEventLogExportManager.swift]: [nit] An observation: I am not a fan of the pattern of returning an `Optional<Error>`.  I would rather use `throws`.  I notice that this pattern appears a lot in this PR.  I won't block it, but my concern is more around future developers following this [anti-]pattern and I don't think it is a good idea.

File: Loop/Views/CriticalEventLogExportViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    private var remainingDuration: TimeInterval?
+    private var backgroundTaskIdentifier: UIBackgroundTaskIdentifier?
+
+    private let log = OSLog(category: "CriticalEventLogExportManager")
+
+    init(exporterFactory: CriticalEventLogExporterFactory) {
+        self.exporterFactory = exporterFactory
+    }
+
+    func export() {
```
Comment (rickpasetto) [inline Loop/Views/CriticalEventLogExportViewModel.swift]: [nit] This code looks a lot more like policy code that maybe belongs in `CriticalEventLogExportManager.swift`.  IMO the `ViewModel` should just talk to that class which does the heavy lifting, and manage updates/events/progress for the view to show.

File: Loop/Views/CriticalEventLogExportViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+    @objc private func willEnterForegroundNotificationReceived(_ notification: Notification) {
+        beginBackgroundTask()
+    }
+
+    // Required due to the current design which creates all settings-related view models at once when the settings view is
+    // about to be displayed. Since it is possible to invoke the export view multiple times from with the same settings view and
+    // this view model contains view-specific state information it is therefore necessary to re-use the same view model each time
+    // the export view thus necessitating the reset functionality.
+    private func reset() {
```
Comment (rickpasetto) [inline Loop/Views/CriticalEventLogExportViewModel.swift]: See my [nit] comment above.  Needing this seems like a code smell to me.  If the responsibilities that need to be `reset()` here were moved to `CriticalEventLogExportManager`, it could (a) allow the ViewModel to be more ephemeral, and (b) put that code under unit test.

File: Loop/AppDelegate.swift [hunk 13]
```diff
@@ -190,6 +194,30 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, DeviceOrientationCo
...
+
+    private func registerCriticalEventLogHistoricalExportBackgroundTask() {
+        if DeviceDataManager.registerCriticalEventLogHistoricalExportBackgroundTask({ self.deviceDataManager.handleCriticalEventLogHistoricalExportBackgroundTask($0) }) {
+            log.debug("Critical event log export background task registered")
+        } else {
+            log.error("Critical event log export background task not registered")
+        }
+    }
+
+    private func scheduleCriticalEventLogHistoricalExportBackgroundTask() {
```
Comment (ps2) [inline Loop/AppDelegate.swift]: [nit] Not sure it's worth creating these one line functions that already have a descriptive name on DDM.

File: Loop/Views/CriticalEventLogExportViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+import SwiftUI
+import LoopKit
+
+protocol CriticalEventLogExporterFactory {
+    func createExporter(to url: URL) -> CriticalEventLogExporter
+}
+
+extension CriticalEventLogExportManager: CriticalEventLogExporterFactory {}
+
+public class CriticalEventLogExportViewModel: ObservableObject, Identifiable, CriticalEventLogExporterDelegate {
```
Comment (rickpasetto) [inline Loop/Views/CriticalEventLogExportViewModel.swift]: Unit test for this class?

File: Loop/Views/CriticalEventLogExportView.swift [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+                .accentColor(Color.loopAccent)
+                .padding()
+            Text(viewModel.remainingDurationString ?? " ")  // Vertical alignment hack
+        }
+    }
+
+    @ViewBuilder
+    private var exportedView: some View {
+        VStack {
+            Image("Checkmark")
```
Comment (ps2) [inline Loop/Views/CriticalEventLogExportView.swift]: Use SFSymbols checkmark.circle.fill here instead?

File: Loop/Managers/ZipArchive.swift [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+import Foundation
+import LoopKit
+import Minizip
+
+public enum ZipArchiveError: Error, Equatable {
+    case unexpectedStatus(Stream.Status)
+    case internalFailure(Int32)
+}
+
+public class ZipArchive {
```
Comment (ps2) [inline Loop/Managers/ZipArchive.swift]: [nit] maybe this goes in Loop/Models instead of Loop/Managers

**Review by ps2 — APPROVED** [review]
A few comments, but overall looks good to me. I'm not sure in which sense these events are "critical", but I don't remember where that term came from; maybe from the iAGC.  Looking at the ticket history, when I created the jira ticket, I called it critical event log. It seems this is maybe more of a "forensic" log? No real recommendation here, I just see that the word "critical" is everywhere, and that seems like a term that implies some highly selective criteria.

File: Loop/Managers/CriticalEventLogExportManager.swift [hunk 17]
```diff
@@ -0,0 +1,513 @@
...
+    public func createHistoricalExporter() -> CriticalEventLogExporter {
+        return CriticalEventLogHistoricalExporter(manager: self)
+    }
+
+    // MARK: - Export synchronization
+
+    private let synchronizeExportExpireDuration: TimeInterval = .seconds(5)
+    private let synchronizeExportCancelDuration: TimeInterval = .seconds(0.1)
+
+    fileprivate func synchronizeExport(for exporter: CriticalEventLogSynchronizedExporter, cancellingActive: Bool, now: Date) -> Error? {
```
Comment (rickpasetto) [inline Loop/Managers/CriticalEventLogExportManager.swift:109]: I'm not exactly sure why this function uses a semaphore.  Why is this needed?  I could use some help understanding it.

**Review by rickpasetto — COMMENTED** [review]
Lots of good stuff, nice work, Darin!  I have some questions I'd like to see addressed before approving.  If you want to hop on a call I'm happy to do it.

File: Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift (hunk 2 above)
Comment (darinkrauss) [reply Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift]: For export, I created a new, Codable `StoredDosingDecision.NotificationSettings` that is stored in Core Data (rather than a binary encoded `UNNotificationSettings`). This simply reflects that change to `StoredDosingDecision`.

Comment (darinkrauss) [reply Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift]: Will pull out into top-level class.

File: Loop/AppDelegate.swift (hunk 13 above)
Comment (darinkrauss) [reply Loop/AppDelegate.swift]: Will remove one set of functions here.

File: Cartfile (hunk 1 above)
Comment (darinkrauss) [reply Cartfile:9]: Per discussion, will leave as-is for now. We can revisit before LMR.

Comment (darinkrauss) [reply Cartfile:9]: https://tidepool.atlassian.net/browse/LOOP-1995

File: Loop/Managers/Alerts/AlertStore.swift (hunk 4 above)
Comment (darinkrauss) [reply Loop/Managers/Alerts/AlertStore.swift:422]: Since historical exports are captured on a daily basis it is necessary to cover the use case of the alert issued during one daily period and the alert acknowledged (or retracted) in a subsequent daily period. If endDate is nil, then the final predicate can simplified slightly at the expense of slightly more complicated code, but that is not the common use case. (Most exports are historical and those will have an endDate specified.) Going to keep as-is.

Regarding an alert that is issued for a future date, but retracted (and deleted) before that future data occurs, it never actually triggers or has user impact, so I'm fine with it not being included in the critical event logs.

File: Loop/Managers/DeviceDataManager.swift (hunk 5 above)
Comment (darinkrauss) [reply Loop/Managers/DeviceDataManager.swift]: Yep, my mistake. Fixed.

File: Loop/Managers/DeviceDataManager.swift (hunk 6 above)
Comment (darinkrauss) [reply Loop/Managers/DeviceDataManager.swift:1057]: Per earlier discussion, going to leave as-is.

File: Loop/Managers/ZipArchive.swift (hunk 7 above)
Comment (darinkrauss) [reply Loop/Managers/ZipArchive.swift]: Upon reviewing @ps's comments, I moved it into the `Model` subgroup.

File: Loop/Managers/CriticalEventLogExportManager.swift (hunk 10 above)
Comment (darinkrauss) [reply Loop/Managers/CriticalEventLogExportManager.swift]: Per earlier discussion, going to leave as-is.

File: Loop/Managers/CriticalEventLogExportManager.swift (hunk 17 above)
Comment (darinkrauss) [reply Loop/Managers/CriticalEventLogExportManager.swift:109]: The semaphore is necessary to coordinate the two exports that could be operating at once. The historical export could be running in the background when the user requests a full export. The semaphore allows the wait for the historical export to cancel before the full export begins. (Running more than one export at a time would be a waste of resources, if not worse.) The semaphore enables this coordination.

**Comment by darinkrauss** [comment]
@ps2 Regarding your concern about the "critical" terminology, I see this term comes directly from iAGC section 4.

 > The device design must ensure that a record of critical events is stored and accessible for an adequate period to allow for

I think the usage of "critical" is appropriate here, even though it may be a slightly overloaded term in the rest of the code.

File: Loop/AppDelegate.swift (hunk 13 above)
Comment (darinkrauss) [reply Loop/AppDelegate.swift]: ✅

File: Loop/Views/CriticalEventLogExportView.swift (hunk 15 above)
Comment (darinkrauss) [reply Loop/Views/CriticalEventLogExportView.swift]: Yep, done. ✅

File: Loop/Managers/ZipArchive.swift (hunk 16 above)
Comment (darinkrauss) [reply Loop/Managers/ZipArchive.swift]: ✅

File: Cartfile (hunk 1 above)
Comment (darinkrauss) [reply Cartfile:9]: ✅

File: Loop/Managers/Alerts/AlertStore.swift (hunk 4 above)
Comment (darinkrauss) [reply Loop/Managers/Alerts/AlertStore.swift:422]: ✅

File: Loop/Managers/DeviceDataManager.swift (hunk 5 above)
Comment (darinkrauss) [reply Loop/Managers/DeviceDataManager.swift]: ✅

File: Loop/Managers/DeviceDataManager.swift (hunk 6 above)
Comment (darinkrauss) [reply Loop/Managers/DeviceDataManager.swift:1057]: ✅

File: Loop/Managers/ZipArchive.swift (hunk 7 above)
Comment (darinkrauss) [reply Loop/Managers/ZipArchive.swift]: ✅

File: Loop/Managers/CriticalEventLogExportManager.swift (hunk 10 above)
Comment (darinkrauss) [reply Loop/Managers/CriticalEventLogExportManager.swift]: ✅

File: Loop/Managers/CriticalEventLogExportManager.swift (hunk 17 above)
Comment (darinkrauss) [reply Loop/Managers/CriticalEventLogExportManager.swift:109]: ✅

File: Loop/Extensions/OutputStream.swift (hunk 3 above)
Comment (darinkrauss) [reply Loop/Extensions/OutputStream.swift:12]: Good catch, done. ✅

File: Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift (hunk 2 above)
Comment (darinkrauss) [reply Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift]: Pulled out into a top-level class in a separate file and added separate unit tests. ✅

File: Loop/Managers/ZipArchive.swift (hunk 8 above)
Comment (darinkrauss) [reply Loop/Managers/ZipArchive.swift]: I refactor this and related code. It should be more understandable and straightforward now. ✅

File: Loop/Managers/ZipArchive.swift (hunk 9 above)
Comment (darinkrauss) [reply Loop/Managers/ZipArchive.swift]: I refactor this and related code. It should be more understandable and straightforward now. ✅

File: Loop/Views/CriticalEventLogExportViewModel.swift (hunk 11 above)
Comment (darinkrauss) [reply Loop/Views/CriticalEventLogExportViewModel.swift]: Refactored most functionality into `CriticalEventLogExportManager.swift`. The view model is now much light. ✅

File: Loop/Views/CriticalEventLogExportViewModel.swift (hunk 12 above)
Comment (darinkrauss) [reply Loop/Views/CriticalEventLogExportViewModel.swift]: Refactored most functionality into `CriticalEventLogExportManager.swift`. The view model is now much light. ✅

File: Loop/Views/CriticalEventLogExportViewModel.swift (hunk 14 above)
Comment (darinkrauss) [reply Loop/Views/CriticalEventLogExportViewModel.swift]: Refactored most functionality into `CriticalEventLogExportManager.swift`. The view model is now much light. ✅

**Review by rickpasetto — APPROVED** [review]
LGTM!  Nice job 🌻
